File: app/translation/google.py
"""Google Cloud Translation API (v2) backend.

Uses the simple API-key REST endpoint (no service account / OAuth needed).
Requests are batched to respect the documented v2 limits — at most 128 strings
and ~30k UTF-8 code points per call — and results are reassembled in order.
"""

from __future__ import annotations

import json
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, urlopen

from app.translation.base import Translator, TranslationError

_ENDPOINT = "https://translation.googleapis.com/language/translate/v2"
# Conservative caps below Google's hard limits (128 segments / 30k code points).
_MAX_SEGMENTS_PER_REQUEST = 100
_MAX_CHARS_PER_REQUEST = 25_000


def _batch(texts: list[str]) -> list[list[str]]:
    """Split ``texts`` into chunks within the per-request segment/char limits.

    A single text longer than the char cap still goes out alone (the API will
    reject it if it truly exceeds the hard limit, surfaced as a TranslationError).
    """
    batches: list[list[str]] = []
    cur: list[str] = []
    cur_chars = 0
    for t in texts:
        n = len(t)
        if cur and (
            len(cur) >= _MAX_SEGMENTS_PER_REQUEST
            or cur_chars + n > _MAX_CHARS_PER_REQUEST
        ):
            batches.append(cur)
            cur, cur_chars = [], 0
        cur.append(t)
        cur_chars += n
    if cur:
        batches.append(cur)
    return batches


class GoogleTranslator(Translator):
    name = "google"
    label = "Google Translate"

    def __init__(self, api_key: str):
        key = (api_key or "").strip()
        if not key:
            raise TranslationError("No Google Translation API key configured.")
        self._key = key

    def translate(
        self,
        texts: list[str],
        target_lang: str,
        source_lang: str | None = None,
    ) -> list[str]:
        if not texts:
            return []
        out: list[str] = []
        for batch in _batch(texts):
            out.extend(self._translate_batch(batch, target_lang, source_lang))
        return out

    def _translate_batch(
        self, batch: list[str], target_lang: str, source_lang: str | None
    ) -> list[str]:
        # Build the body as repeated q= fields (the v2 API accepts q arrays).
        params: list[tuple[str, str]] = [("target", target_lang), ("format", "text")]
        if source_lang:
            params.append(("source", source_lang))
        params.extend(("q", t) for t in batch)
        body = urlencode(params).encode("utf-8")
        url = f"{_ENDPOINT}?{urlencode({'key': self._key})}"
        req = Request(url, data=body, method="POST")  # noqa: S310 - fixed https host
        req.add_header("Content-Type", "application/x-www-form-urlencoded")
        try:
            with urlopen(req, timeout=60) as resp:  # noqa: S310 - fixed https host
                payload = json.loads(resp.read())
        except HTTPError as exc:
            detail = exc.read().decode("utf-8", "replace") if exc.fp else str(exc)
            raise TranslationError(
                f"Google Translation API error ({exc.code}): {detail}"
            ) from exc
        except URLError as exc:
            raise TranslationError(f"Couldn't reach Google: {exc.reason}") from exc

        translations = (payload.get("data") or {}).get("translations") or []
        if len(translations) != len(batch):
            raise TranslationError(
                "Unexpected response from Google: "
                f"{len(translations)} translations for {len(batch)} inputs."
            )
        return [t.get("translatedText", "") for t in translations]

File: adapters/server/translate/google.cpp
#include "translate/google.hpp"

#include <nlohmann/json.hpp>

#include "http/curl_client.hpp"

namespace whisperx::server::translate {

namespace {

using nlohmann::json;
namespace net = whisperx::server::net;

constexpr const char* kEndpoint =
    "https://translation.googleapis.com/language/translate/v2";
// Conservative caps below Google's hard limits (128 segments / 30k code points).
constexpr std::size_t kMaxSegmentsPerRequest = 100;
constexpr std::size_t kMaxCharsPerRequest = 25000;

// Split `texts` into chunks within the per-request segment/char limits. A single
// text longer than the char cap still goes out alone (the API rejects it if it
// truly exceeds the hard limit, surfaced as a TranslationError).
std::vector<std::vector<std::string>> batch(
    const std::vector<std::string>& texts) {
    std::vector<std::vector<std::string>> batches;
    std::vector<std::string> cur;
    std::size_t cur_chars = 0;
    for (const auto& t : texts) {
        if (!cur.empty() && (cur.size() >= kMaxSegmentsPerRequest ||
                             cur_chars + t.size() > kMaxCharsPerRequest)) {
            batches.push_back(std::move(cur));
            cur.clear();
            cur_chars = 0;
        }
        cur_chars += t.size();
        cur.push_back(t);
    }
    if (!cur.empty()) batches.push_back(std::move(cur));
    return batches;
}

std::vector<std::string> translate_batch(const std::vector<std::string>& b,
                                         const std::string& target_lang,
                                         const std::string& api_key,
                                         const std::string& source_lang) {
    // Body as repeated q= fields (the v2 API accepts q arrays).
    std::vector<std::pair<std::string, std::string>> fields = {
        {"target", target_lang}, {"format", "text"}};
    if (!source_lang.empty()) fields.emplace_back("source", source_lang);
    for (const auto& t : b) fields.emplace_back("q", t);

    std::string url = std::string(kEndpoint) + "?" +
                      net::form_encode({{"key", api_key}});
    net::Response resp = net::post(url, net::form_encode(fields), {{}, 60, true});

    if (resp.status == 0)
        throw TranslationError("Couldn't reach Google: request failed");
    if (!resp.ok())
        throw TranslationError("Google Translation API error (" +
                               std::to_string(resp.status) + "): " + resp.body);

    json payload = json::parse(resp.body, nullptr, false);
    if (payload.is_discarded())
        throw TranslationError("Unexpected response from Google: invalid JSON.");
    const json& translations =
        payload.value("data", json::object()).value("translations", json::array());
    if (translations.size() != b.size())
        throw TranslationError(
            "Unexpected response from Google: " +
            std::to_string(translations.size()) + " translations for " +
            std::to_string(b.size()) + " inputs.");
    std::vector<std::string> out;
    out.reserve(b.size());
    for (const auto& t : translations)
        out.push_back(t.value("translatedText", std::string()));
    return out;
}

}  // namespace

std::vector<std::string> google_translate(
    const std::vector<std::string>& texts, const std::string& target_lang,
    const std::string& api_key, const std::string& source_lang) {
    if (api_key.empty())
        throw TranslationError("No Google Translation API key configured.");
    if (texts.empty()) return {};
    std::vector<std::string> out;
    out.reserve(texts.size());
    for (const auto& b : batch(texts)) {
        auto part = translate_batch(b, target_lang, api_key, source_lang);
        out.insert(out.end(), part.begin(), part.end());
    }
    return out;
}

std::pair<bool, std::string> verify_api_key(const std::string& key) {
    // trim
    auto b = key.find_first_not_of(" \t\r\n");
    std::string k = (b == std::string::npos)
                        ? ""
                        : key.substr(b, key.find_last_not_of(" \t\r\n") - b + 1);
    if (k.empty()) return {false, "Enter an API key to continue."};

    // Hits the (free, cheap) supported-languages endpoint to confirm the key is
    // valid and the Translation API is enabled for it.
    std::string url = std::string(kEndpoint) + "/languages?" +
                      net::form_encode({{"key", k}});
    net::Response resp = net::get(url, {{}, 15, true});
    if (resp.status == 0) return {false, "Couldn't reach Google: request failed"};
    if (resp.status == 400 || resp.status == 401 || resp.status == 403)
        return {false,
                "Invalid or unauthorized API key. Check the key and that the "
                "Cloud Translation API is enabled for its project."};
    if (!resp.ok())
        return {false, "Couldn't verify key: HTTP " + std::to_string(resp.status)};
    return {true, "Key verified — translation is ready."};
}

}  // namespace whisperx::server::translate
